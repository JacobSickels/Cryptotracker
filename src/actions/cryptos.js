import database from '../firebase/firebase';

// SET_CRYPTOS
export const setCryptos = (cryptos) => ({
    type: 'SET_CRYPTOS',
    cryptos
});


// SET_START_CRYPTOS
// This is a dispatching wrapper for the setCryptos action object
// This method goes to firebase, grabs the necessary data, and loads an object to be passed and dispatched to the redux store.

export const startSetCryptos = () => {
    return (dispatch) => {
        return database.ref('crypto_data')
            .once('value')
            .then((snapshot) => {

                // Creating an object to load Firebase data into
                const cryptos = {
                    bitcoin: [],
                    litecoin: [],
                    ethereum: []
                };

                // Looping through Firebase data and loading the object
                snapshot.forEach((childSnapshot) => {
                    
                    if(childSnapshot.key === 'bitcoin') {
                        Object.keys(childSnapshot.val()).forEach(function(key) {
                            cryptos.bitcoin.push(childSnapshot.val()[key]);
                        });
                    }
                    else if(childSnapshot.key === 'litecoin') {
                        Object.keys(childSnapshot.val()).forEach(function(key) {
                            cryptos.litecoin.push(childSnapshot.val()[key]);
                        });
                    }
                    else {
                        Object.keys(childSnapshot.val()).forEach(function(key) {
                            cryptos.ethereum.push(childSnapshot.val()[key]);
                        });
                    }

                });

                // Dispatching object to the Redux store.
                dispatch(setCryptos(cryptos));

            });
    };
}