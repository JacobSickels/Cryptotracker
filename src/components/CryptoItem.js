import React from 'react';
import { Link } from 'react-router-dom'; //Eventually linking to the specific cryptos page for more specific filtering
import moment from 'moment';
import { connect } from 'react-redux';
import { ResponsiveContainer, LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend } from 'recharts';
import selectCryptos from '../selectors/cryptos';

export class CryptoItem extends React.Component {
    
    constructor(props) {
        super(props);

        this.state = {
            showTrendLine: true
        }
    }

    render() {
        return (
            <div className="chart">
                <div className="chart-title">
                    <div className="container">
                        <div className="chart-title--container">
                            <div>
                                <Link className="chart-title__link" to={`/filter/${this.props.name.toLowerCase()}`} >
                                    <h1> {this.props.name} </h1>
                                </Link>
                            </div>
                            <div className="chart-title__amount">
                                <h1>
                                    {this.props.currency_symbol} {this.props.current.amount}
                                </h1>
                            </div>
                            <div className="chart-title--percentage">
                                 <h1 style={{
                                     color:
                                        (this.props.percentage >= 0) ? 'green' : 'red'
                                     }}>
                                    {this.props.percentage.toFixed(2)}% 
                                </h1>
                            </div>
                        </div>
                    </div>
                </div>
                <div className="container">
                    <div className="button-container">
                        <button className="button--trend"
                            onClick={() => {
                                const show = this.state.showTrendLine;
                                this.setState({showTrendLine: !show})
                            }}>
                        {
                            (this.state.showTrendLine) ? 'Hide Trend' : 'Show Trend'
                        }      
                        </button>
                    </div>
                    <div className="chart-container">
                        <ResponsiveContainer height='100%' width='100%'>
                            <LineChart data={this.props.data} margin={{top: 5, right: 10, left: 30, bottom: 5}}>
                                <XAxis dataKey="timestamp"/>
                                <YAxis tickSize={10} dataKey="amount" domain={["dataMin", "dataMax"]} allowDataOverflow={true} />
                                <CartesianGrid strokeDasharray="3 3" stroke="#FFDDED"/>
                                <Tooltip/>
                                <Legend />
                                {
                                    (this.state.showTrendLine) && (
                                        <Line dot={false} type="monotone" dataKey="trend" stroke="#E477D4" strokeWidth={4}/>
                                    )
                                }
                                <Line dot={false} type="monotone" dataKey="amount" stroke="#7F4276"/>
                            </LineChart>
                        </ResponsiveContainer>
                    </div>
                </div>
            </div>
        );
    };
}

const convertToExchange = (exchangeRate, cryptos) => {
    const converted = cryptos.map((entry) => ({
        amount: (parseFloat(entry.amount) * exchangeRate).toFixed(2),
        timestamp: entry.timestamp
    }));

    return converted;
}

const getTrendData = (crypto) => {
    let x = [];
    
    for(let i = 0; i < crypto.length; i++) {
        x.push(i);
    }

    const xSum = x.map(num => num).reduce((prev, next) => prev + next);
    const ySum = crypto.map(entry => parseFloat(entry.amount)).reduce((prev, next) => prev + next);
    const averageX = xSum / crypto.length;
    const averageY = ySum / crypto.length;

    const topSlopeArray = [];
    for(let i = 0; i < crypto.length; i++) {
        topSlopeArray.push((x[i] - averageX) * (crypto[i].amount - averageY));
    }
    const topSlope = topSlopeArray.reduce((prev, next) => prev + next);
    const bottomSlope = x.map(num => (num - averageX) * (num - averageX)).reduce ((prev, next) => prev + next);

    const slope = topSlope / bottomSlope;
    const yIntercept = averageY - (slope * averageX);
    const percentage = (crypto[crypto.length - 1].amount - crypto[0].amount) / crypto[0].amount * 100;

    const data = crypto.map((entry, index) => ({
                timestamp: moment(entry.timestamp).format('ha'), 
                amount: parseFloat(entry.amount), 
                trend: slope * index + yIntercept
            }));

    return {
        data,
        percentage,
        current: crypto[crypto.length - 1]
    };
};

const mapStateToProps = (state, props) => {

    let crypto = selectCryptos(state.cryptos, state.filters)[props.name.toLowerCase()];

    const exchangeRate = parseFloat(state.filters.currency.exchange_rate);

    //converting amounts to exchange rate
    crypto = convertToExchange(exchangeRate, crypto);

    return {
        ...getTrendData(crypto),
        currency_symbol: state.filters.currency.symbol
    }
    
};

export default connect(mapStateToProps)(CryptoItem);