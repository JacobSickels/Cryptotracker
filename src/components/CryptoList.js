import React from 'react';
import { connect } from 'react-redux';
import selectCryptos from '../selectors/cryptos';

import CryptoListItem from './CryptoListItem';

export const CryptoList = (props) => (
    <div> 
        <h2>CryptoList</h2> 
        Bitcoin:
        <CryptoListItem crypto={props.cryptos.bitcoin} />
        Litecoin:
        <CryptoListItem crypto={props.cryptos.litecoin} /> 
        Ethereum:
        <CryptoListItem crypto={props.cryptos.ethereum} />
    </div>
);

const mapStateToProps = (state) => {
    return {
        cryptos: selectCryptos(state.cryptos, state.filters)
    };
};

export default connect(mapStateToProps)(CryptoList);