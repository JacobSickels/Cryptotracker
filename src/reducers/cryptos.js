/*
    This is the Cryptos Data Reducer
    This deals with dispatch actions to cryptos section of Redux state.
*/

// This is the default store state for all of the crypto data
const cryptosReducerDefaultState = {
    bitcoin: [],
    litecoin: [],
    ethereum: []
};

/*

    This function is the reducer
    action.type(s) come into the function, and the reducer updates the state accordingly

*/
export default (state = cryptosReducerDefaultState, action) => {

    switch(action.type) {
        //This updates state to action.cryptos
        case 'SET_CRYPTOS':
            return action.cryptos;
        //When there is no action, the current state is returned    
        default: return state;
    }

};