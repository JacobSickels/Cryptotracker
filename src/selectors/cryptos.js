import moment from 'moment';

//filtering crypto data by date ranges

export default (cryptos, {startDate, endDate}) => {
    const filteredCryptos = {
        bitcoin: [],
        litecoin: [],
        ethereum: []
    };
    for(var prop in cryptos) {
        filteredCryptos[prop] = cryptos[prop].filter((crypto) => {
            //console.log('checking crypto object',crypto);
            const createdAtMoment = moment(crypto.timestamp);
            const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
            const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

            return startDateMatch && endDateMatch;
        });

        //Fixes issue on Chrome
        filteredCryptos[prop].sort((a, b) => {
            return a.timestamp - b.timestamp;
        });

    }

    return filteredCryptos;
};
