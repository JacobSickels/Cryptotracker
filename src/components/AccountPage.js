import React from 'react';
import { connect } from 'react-redux';
import { Input } from 'react-materialize';

import { startEditCurrency } from '../actions/filters';

export class AccountPage extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            currency: props.currency
        }

    }

    onInputChange = (e) => {
        const currency = this.props.exchanges.find((element) => element.id === e.target.value);
        this.setState({ currency });
        this.props.startEditCurrency(currency);
    }

    render() {
        return (
            <div className="page-container">
                <div className="container">
                    <h2>My Account Page</h2>

                    <h1>Set Base Currency</h1>
                    <Input s={5} type='select' label="Select Currency" value={this.state.currency.id} onChange={this.onInputChange}>
                        {
                            this.props.exchanges.map((currency) =>
                                <option key={currency.id} value={currency.id}>{currency.name}</option>
                            )
                        }
                    </Input>
                </div>
            </div>
        );
    }

}

const mapDispatchToProps = (dispatch) => {
    return {
        startEditCurrency: (currency) => dispatch(startEditCurrency(currency))
    };
};

const mapStateToProps = (state, props) => {
    return {
        exchanges: state.exchange,
        currency: state.filters.currency
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(AccountPage);