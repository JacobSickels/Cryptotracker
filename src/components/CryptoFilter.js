import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';

import { setStartDate, setEndDate } from '../actions/filters';

export class CryptoFilter extends React.Component {

    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    render() {
        return (
            <div className="date-container">
                <div className="container">
                    <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        minimumNights={0}
                        isOutsideRange={(day) => {
                            const today = moment();
                            return day >= today;
                        }}
                    />
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(CryptoFilter);