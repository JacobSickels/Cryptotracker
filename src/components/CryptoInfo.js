import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { Row, Col } from 'react-materialize';

import { selectCryptos, getMaxMinData } from '../selectors/cryptos';

export class CryptoInfo extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            maxDifference: () => {
                const crypto = props.data;
    
                let maxDifference = 0;
                let minElement = crypto[0].amount;
                
                for(var i = 1; i < crypto.length; i++) {
                    if(crypto[i].amount - minElement > maxDifference) {
                        maxDifference = crypto[i].amount - minElement;
                    }
                    if(crypto[i] < minElement) {
                        minElement = crypto[i].amount;
                    }
                }
                return maxDifference.toFixed(2);
            }
        }
    }

    render() {
        return (
            <div className="chart-info">
                <div className="container">
                    <Row> 
                        <Col s={12} m={4} l={4}>
                            <h1> Max Increase </h1>
                            {this.props.maxDifference}
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Buy</h1> 
                            <em>{moment(this.props.minElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.minElement.amount}</p>
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Sell</h1>
                            <em>{moment(this.props.maxElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.maxElement.amount}</p>
                        </Col>
                    </Row>
                </div>
            </div>
        );
    }

}

const mapStateToProps = (state, props) => {
    const data = getMaxMinData(state.cryptos, state.filters, props.name.toLowerCase());
    return { ...data }
};

export default connect(mapStateToProps)(CryptoInfo);