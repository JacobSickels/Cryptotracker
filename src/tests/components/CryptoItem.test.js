import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import { CryptoItem } from '../../components/CryptoItem';
import cryptos from '../fixtures/cryptos';
import getTrendData from '../../selectors/trend';

//This test makes sure that CryptoItem is rendered correctly
test('Should render CryptoItem correctly', () => {

    const currency = {
        "exchange_rate": "69.50",
        "id": "AFN",
        "name": "Afghan Afghani",
        "symbol": "Ø‹"
    };

    const trendData = getTrendData(cryptos.bitcoin);

    const wrapper = shallow(
    <CryptoItem 
        name="bitcoin"
        currency_symbol={currency.symbol}
        current={trendData.current}
        percentage={trendData.percentage}
        data={trendData.data}
    />);
    expect(wrapper).toMatchSnapshot();
});