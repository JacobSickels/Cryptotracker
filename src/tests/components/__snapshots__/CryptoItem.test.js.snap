// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render CryptoItem correctly 1`] = `
<div
  className="chart"
>
  <div
    className="chart-title"
  >
    <div
      className="container"
    >
      <div
        className="chart-title--container"
      >
        <div>
          <Link
            className="chart-title__link"
            replace={false}
            to="/filter/bitcoin"
          >
            <h1>
               
              bitcoin
               
            </h1>
          </Link>
        </div>
        <div
          className="chart-title__amount"
        >
          <h1>
            Ø‹
             
            222222
          </h1>
        </div>
        <div
          className="chart-title--percentage"
        >
          <h1
            style={
              Object {
                "color": "green",
              }
            }
          >
            80.00
            %
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="button-container"
    >
      <button
        className="button--trend"
        onClick={[Function]}
      >
        Hide Trend
      </button>
    </div>
    <div
      className="chart-container"
    >
      <ResponsiveContainer
        debounce={0}
        height="100%"
        width="100%"
      >
        <LineChart
          barCategoryGap="10%"
          barGap={4}
          data={
            Array [
              Object {
                "amount": 123456,
                "timestamp": "6pm",
                "trend": 123456,
              },
              Object {
                "amount": 222222,
                "timestamp": "6pm",
                "trend": 222222,
              },
            ]
          }
          layout="horizontal"
          margin={
            Object {
              "bottom": 5,
              "left": 30,
              "right": 10,
              "top": 5,
            }
          }
          reverseStackOrder={false}
          stackOffset="none"
        >
          <XAxis
            allowDataOverflow={false}
            allowDecimals={true}
            allowDuplicatedCategory={true}
            dataKey="timestamp"
            domain={
              Array [
                0,
                "auto",
              ]
            }
            height={30}
            hide={false}
            mirror={false}
            orientation="bottom"
            padding={
              Object {
                "left": 0,
                "right": 0,
              }
            }
            reversed={false}
            scale="auto"
            tickCount={5}
            type="category"
            width={0}
            xAxisId={0}
          />
          <YAxis
            allowDataOverflow={true}
            allowDecimals={true}
            allowDuplicatedCategory={true}
            dataKey="amount"
            domain={
              Array [
                "dataMin",
                "dataMax",
              ]
            }
            height={0}
            hide={false}
            mirror={false}
            orientation="left"
            padding={
              Object {
                "bottom": 0,
                "top": 0,
              }
            }
            reversed={false}
            scale="auto"
            tickCount={5}
            tickSize={10}
            type="number"
            width={60}
            yAxisId={0}
          />
          <CartesianGrid
            fill="none"
            horizontal={true}
            horizontalFill={Array []}
            horizontalPoints={Array []}
            stroke="#FFDDED"
            strokeDasharray="3 3"
            vertical={true}
            verticalFill={Array []}
            verticalPoints={Array []}
          />
          <Tooltip
            active={false}
            animationDuration={400}
            animationEasing="ease"
            coordinate={
              Object {
                "x": 0,
                "y": 0,
              }
            }
            cursor={true}
            cursorStyle={Object {}}
            filterNull={true}
            isAnimationActive={true}
            itemSorter={[Function]}
            itemStyle={Object {}}
            labelStyle={Object {}}
            offset={10}
            separator=" : "
            useTranslate3d={false}
            viewBox={
              Object {
                "x1": 0,
                "x2": 0,
                "y1": 0,
                "y2": 0,
              }
            }
            wrapperStyle={Object {}}
          />
          <Legend
            align="center"
            iconSize={14}
            layout="horizontal"
            verticalAlign="bottom"
          />
          <Line
            activeDot={true}
            animationBegin={0}
            animationDuration={1500}
            animationEasing="ease"
            connectNulls={false}
            dataKey="trend"
            dot={false}
            fill="#fff"
            hide={false}
            isAnimationActive={true}
            legendType="line"
            onAnimationEnd={[Function]}
            onAnimationStart={[Function]}
            points={Array []}
            stroke="#E477D4"
            strokeWidth={4}
            type="monotone"
            xAxisId={0}
            yAxisId={0}
          />
          <Line
            activeDot={true}
            animationBegin={0}
            animationDuration={1500}
            animationEasing="ease"
            connectNulls={false}
            dataKey="amount"
            dot={false}
            fill="#fff"
            hide={false}
            isAnimationActive={true}
            legendType="line"
            onAnimationEnd={[Function]}
            onAnimationStart={[Function]}
            points={Array []}
            stroke="#7F4276"
            strokeWidth={1}
            type="monotone"
            xAxisId={0}
            yAxisId={0}
          />
        </LineChart>
      </ResponsiveContainer>
    </div>
  </div>
</div>
`;
